Remote Sync

johnswanson opened the pull request:
## Project Info
### References
[Linear Project](https://linear.app/metabase/project/remote-sync-3adfe44791dd/overview)
[Shipping Goals](https://linear.app/metabase/document/shipping-goals-remote-sync-338379809bbe)

### Why

We have 2 main motivations for this project:

1. The current process of moving content from one Metabase instance to another is cumbersome (see [documentation](https://www.metabase.com/learn/metabase-basics/administration/administration-and-operation/git-based-workflow)). This project allows users to migrate Collections without having to work outside of the Metabase application.
2. This is a building block towards the Library, which will use this functionality to store important content in a secure external location.

### What’s in

**New Admin Settings**
- A “Remote Sync” section in Admin settings
- Configurable settings for a Git repository, and an access token
- A way to select Development or Production mode for synced content

**Synced Collection UX - Development Mode**
- In Development mode, users can see your synced collection under a new menu in the sidenav
- When changes are made to the synced collection, there are visual indicators to show that changes have been made
- Admins users have the ability to switch branches in the branch picker, or create new branches
- Admins can push new changes to the selected branch

**Synced Collection UX - Production Mode**
- In Production mode, the synced collection is visible in the sidenav with a special icon
- The synced collection and its contents **cannot** be edited in any way in Production mode

**Synced collections are fully self-contained**
- A collection inside the synced collection cannot have references to any content outside of itself, except for other collections that are synced
- If a user tries to sync a collection that has outside references, they will see an error explaining why they cannot

**Pulling new changes**
- If auto-sync is enabled, new changes pushed to Git will be imported automatically after 5 minutes
- Users can also manually trigger a pull from Admin settings

**Permissions**
- Only Admins can push changes to Git

**File Format**
- Collections pushed to Git use existing serialization format

edpaget commented on src/metabase/lib/database.cljc:
```diff
@@ -15,7 +14,8 @@
   we will attempt to resolve the correct Database ID by getting metadata for the source Card and returning its
   `:database-id`; if this is not available for one reason or another this will return `nil`."
-  [query :- ::lib.schema/query]
+  ;; TODO(johnswanson) wtf did I have to remove the query schema bit here
```
@johnswanson what was up here?

edpaget commented:
From a pair review with @nvoxland need to 

> Gotta check dashboard dependencies for the model attached to an action
> 
> Potentially make the dependent check a union-all
> 
> non-remote-synced-dependencies needs to check links from documents and dashboards too
> 
> make non-check version of the dep checking private
> 
> dashboards need check dependents in dashboard api

Main thing was realizing dashboards/documents can have links and we need to follow those references too.

edpaget committed:
Fix remote sync settings test

edpaget committed:
Fix sync type setting validation test

edpaget committed:
Fix check sync task type is production

edpaget committed:
Update plain type to collection_type

edpaget committed:
Cascade changes to collection type

edpaget committed:
Update deps checks when collection type changes

edpaget committed:
Fix task type check in export

edpaget committed:
move setting check to api for better status report

edpaget committed:
add tests for sync-type

nvoxland committed:
Don't mark newly created entities as updated or removed UXW-1806

alxnddr committed:
UXW-1821 enabled badge

alxnddr committed:
UXW-1818: hide branch picker and push button for non-admins

alxnddr committed:
cleanup and UXW-1814

alxnddr committed:
remove border from the branch picker button

alxnddr committed:
cleanup

alxnddr committed:
UXW-1829 update diff modal ui

alxnddr committed:
UXW-1807: token should appear masked

nvoxland committed:
Allow setting of nil token

edpaget committed:
Add new collection update/create events

edpaget committed:
Remove per collection imp/exp; change dirty status table

alxnddr committed:
UXW-1833: show backend error when saving invalid git settings

edpaget committed:
Redo event handling add removed status

edpaget committed:
Sync object table on syncs

edpaget committed:
Fix status names

nvoxland committed:
Added /current-task/cancel API and track version in remote_sync_task

edpaget committed:
Add is_remote_synced to api responses

edpaget committed:
Fix adding is_remote_sync hydration

edpaget committed:
Merge branch 'master' into source-of-truth

nvoxland committed:
Fixed compilation error

edpaget committed:
remove hashp

edpaget committed:
Fix loading is_remote_sync hydration

edpaget committed:
Resolve ambiguous reference

johnswanson committed:
Fix n+1 detector

if it gets a heterogeneous set of things we should hydrate each type
separately.

edpaget committed:
Add can-create? handlers

johnswanson committed:
remote sync: a few bugfixes and improvements from my bugbash (#64372)

* Set `collection.type` on archive/unarchive

When unarchiving or archiving collections that have the `remote-synced`
type, we need to set their new type appropriately.

When archiving, we should give them a normal (nil) type.

When unarchiving any collection *to* a remote-synced collection, we
should set the type to `remote-synced`.

* Allow serdes without archived items

and use it in remote sync export.

During dependency checking, we do *not* want to exclude archived items:
if card A uses card B, and card B is archived, we need to include Card B
in the list of dependencies! Otherwise we'll say that it's fine to
remote sync card A when it's not.

* Add tests

* Documents are descendants of collections

* don't include cards as document descendants

* fix tests, dashboard descendants args

* remove #p and unnecessary query

* remove another `#p`

edpaget committed:
Update auto sync token to enable import job

edpaget committed:
More is_remote_sync hydration fixes

edpaget committed:
Fix lints and document oss serdes

alxnddr committed:
UXW-1817: change branches should update content

edpaget committed:
Fix cards to be valid for update events

nvoxland committed:
Don't actually import if nothing has changed in the remote repo

edpaget committed:
Remove hashp

nvoxland committed:
Merge remote-tracking branch 'origin/source-of-truth' into source-of-truth

edpaget committed:
Fix ee-available? check

alxnddr committed:
UXW-1813 ensure sync dirty state is properly invalidated

alxnddr committed:
UXW-1864 fix how collections changes appear in the diff

edpaget committed:
Improve is_remote_synced hydration

edpaget committed:
Merge branch 'master' into source-of-truth

edpaget committed:
Fix some more oss/ee issues

finish remote ChangeLog model

alxnddr committed:
fix invalidation on removal

edpaget committed:
fix missing require

nvoxland committed:
Don't need to handle transforms directories specially

nvoxland committed:
Merge remote-tracking branch 'origin/source-of-truth' into source-of-truth

edpaget committed:
fix tests to make sure state is fully reset

edpaget committed:
remove duplicate require

edpaget committed:
Fix deadlock on import

edpaget committed:
Fixing tests

edpaget committed:
Fix document tests

edpaget committed:
Fix kondos

edpaget committed:
Merge branch 'master' into source-of-truth

alxnddr committed:
handle removed sync_status

alxnddr committed:
UXW-1851: show moving out of synced collection error

alxnddr committed:
feature flag, misc fixes

edpaget committed:
dirty state tracking for snippets

edpaget committed:
Fix ns typo

edpaget committed:
Add remote-sync token and config

alxnddr committed:
Merge branch 'master' into source-of-truth

alxnddr committed:
enforce remote_sync token feature on the FE

andsalves committed:
allow cancelling the remote-sync

nvoxland committed:
Track timeline

nvoxland committed:
Merge remote-tracking branch 'origin/source-of-truth' into source-of-truth

nvoxland committed:
Force re-import when configuring for production mode

nvoxland committed:
Fixed compilation error

nvoxland committed:
Force re-import when configuring for production mode only when auto-sync is enabled

alxnddr committed:
force import button

nvoxland committed:
Throwa 400 error if there are dirty objects when doing an import

edpaget committed:
Better archiving checks with remote synced collections

edpaget committed:
Merge branch 'source-of-truth-can-create' into source-of-truth

edpaget committed:
Better tests for documents and remote-sync deps

edpaget committed:
Remove hashp

edpaget committed:
remote hashp

edpaget committed:
Remove other hashp

edpaget committed:
More consistent collection write checks

edpaget committed:
Fix broken collection tests

alxnddr committed:
improve imports/exports frontend

edpaget committed:
Fix broken documents tests

edpaget committed:
Implement can-create? for timeline-event

alxnddr committed:
tweak copy

edpaget committed:
Always update branch setting

nvoxland committed:
Initial Commit

edpaget committed:
Fix more tests

edpaget committed:
Remove bad parallel annotations

andsalves committed:
Show error toast when adding a collection that has references outside of the repo

alxnddr committed:
reasonable structure, error handling

alxnddr committed:
timelines editing is less frequent, we can invalidate dirty state on any edit

edpaget committed:
Fix dirty state api tests for mysql

alxnddr committed:
bring error details back

edpaget committed:
Fix collection move tests

edpaget committed:
Fix remove parallel from can_restore tests

Splitting this into smaller tests seems to introduce some kind of state
that causes tests to fail in parallel

edpaget committed:
Fix make logs use unique names for be tests

edpaget committed:
Bump toucan2 versions

edpaget committed:
Fix var name

edpaget committed:
Fix check that there are exportable collections

nvoxland committed:
Merge remote-tracking branch 'origin/source-of-truth' into source-of-truth

edpaget committed:
Add test for exportable check

nvoxland committed:
Started on e2e test

nvoxland committed:
Merge remote-tracking branch 'origin/source-of-truth' into source-of-truth

edpaget committed:
Merge branch 'master' into source-of-truth

edpaget committed:
Fix issue with merge

edpaget committed:
Fix arguments to card descendants

edpaget committed:
Handle new ingestable interface

nvoxland committed:
Put back changes in cypress-runner

nvoxland committed:
Merge remote-tracking branch 'origin/source-of-truth' into source-of-truth

edpaget committed:
Fix remaining failing enterprise tests

edpaget committed:
Fix one set of app tests

edpaget committed:
Clean remote sync state in obj tests

alxnddr committed:
single collection

andsalves committed:
Fix unit test - enterprise getIcon function only applied when new feature is enabled (ResultItem.unit.spec.tsx)

andsalves committed:
Fix ts issues / Clean up

alxnddr committed:
FE fixes

edpaget committed:
Merge branch 'master' into source-of-truth

edpaget committed:
Remove parallel annotations

edpaget committed:
Merge branch 'master' into source-of-truth

edpaget committed:
Fix missing import from merging master

edpaget committed:
Fix the fix

alxnddr committed:
always create collection to sync

alxnddr committed:
no unsyncing

edpaget committed:
Remove associating location to trash from search

edpaget committed:
Merge branch 'master' into source-of-truth

andsalves committed:
Prevent users from switching branches with unsynced changes + show warning

nvoxland committed:
Create remote-sync collection when it doesn't exist

andsalves committed:
Prevent changing to production with unsynced changes

alxnddr committed:
single synced collection ui update

andsalves committed:
Make sure dirty state refetches on admin settings load

nvoxland committed:
Cannot change to production mode when there are unsynced changes

nvoxland committed:
Merge remote-tracking branch 'origin/source-of-truth' into source-of-truth

edpaget committed:
Fix failing query

edpaget committed:
unquote branches

edpaget committed:
fix eastwood

edpaget committed:
fix timeline event publishing

edpaget committed:
fix import

edpaget committed:
Move dashboard remote sync api tests into ee

...where they can reference the setting

nvoxland committed:
Created /stash API and renamed create branch api to /create-branch

nvoxland committed:
Merge remote-tracking branch 'origin/source-of-truth' into source-of-truth

# Conflicts:
#	enterprise/backend/test/metabase_enterprise/remote_sync/api_test.clj

edpaget committed:
Fix settings issue

edpaget committed:
Fix remaining permissions message tests

edpaget committed:
Make sure drop target always works for trash

edpaget committed:
Merge branch 'master' into source-of-truth

edpaget committed:
Add checks for if we're in trash elsewhere

alxnddr committed:
cleanup

alxnddr committed:
git-sync to remote-sync

alxnddr committed:
more git-sync to remote-sync

nvoxland committed:
Merge remote-tracking branch 'origin/source-of-truth' into source-of-truth

# Conflicts:
#	enterprise/frontend/src/metabase-enterprise/api/remote-sync.ts

nvoxland committed:
Don't allow export if it overwrites an external commit

nvoxland committed:
- Don't assume a "main" branch
- Attempt an import when no remote-sync branch exists when enabling dev mode

nvoxland committed:
Fix creating remote sync library

alxnddr committed:
refactor to middleware

alxnddr committed:
hack: add a delay for refetching instance settings or dirty state after remote sync task finishes

alxnddr committed:
UXW-1953: confirmation flow when switching branches in prod

eliotmetabase committed:
Merge branch 'master' into source-of-truth

andsalves committed:
Update modal when switching branches with dirty state / Component refactorings

npfitz commented on enterprise/frontend/src/metabase-enterprise/remote_sync/components/ChangesLists/AllChangesView.tsx:
```diff
@@ -0,0 +1,155 @@
+                <Group
+                  p="sm"
+                  gap="sm"
+                  mb={group.items.length > 0 ? "12px" : 0}
```
prefer rem

npfitz commented on enterprise/frontend/src/metabase-enterprise/remote_sync/components/ChangesLists/CollectionPath.tsx:
```diff
@@ -0,0 +1,38 @@
+  segments: CollectionPathSegment[];
+}
+
+export const CollectionPath = ({ segments }: CollectionPathProps) => {
```
Did we try using one of the existing `CollectionBreadcrumb` components? If not, it might be worth checking out rather than making another one

npfitz commented on enterprise/frontend/src/metabase-enterprise/remote_sync/components/ChangesLists/index.tsx:
```diff
@@ -0,0 +1,48 @@
+  title?: string;
+}
+
+export const ChangesLists = ({ collections, title }: ChangesListsProps) => {
```
Maybe this is a new convention we're following, but normally our index files just re-export things

npfitz commented on enterprise/frontend/src/metabase-enterprise/remote_sync/components/CollectionSyncManager/CollectionSyncManager.module.css:
```diff
@@ -0,0 +1,15 @@
+.CollectionSelectFolderIcon {
```
Is this file still imported?

npfitz commented on enterprise/frontend/src/metabase-enterprise/remote_sync/components/PushChangesModal/PushChangesModal.tsx:
```diff
@@ -0,0 +1,140 @@
+
+      <Box px="xl" pb="lg">
+        {hasConflict && (
+          <Card bg="warning-light" p="sm" mb="md">
```
Should this be an `Alert`?

npfitz commented on enterprise/frontend/src/metabase-enterprise/remote_sync/utils.ts:
```diff
@@ -0,0 +1,209 @@
+export const getSyncStatusColor = (status: RemoteSyncEntityStatus): string => {
+  switch (status) {
+    case "create":
+      return "var(--mb-base-color-palm-50)";
```
I don't think we're supposed to reference these directly

npfitz commented on frontend/src/metabase/collections/components/CollectionHeader/CollectionCaption.tsx:
```diff
@@ -92,6 +92,10 @@ const CollectionCaptionIcon = ({ collection }: { collection: Collection }) => {
     );
   }
 
+  if (PLUGIN_COLLECTIONS.isSyncedCollection(collection)) {
```
I find it weird that synced collections and instance analytics collection checks are done so differently

nvoxland committed:
Create a branch off the last-used commit, not the branch itself

npfitz commented on frontend/src/metabase/collections/containers/FormCollectionAndDashboardPicker/FormCollectionAndDashboardPicker.tsx:
```diff
@@ -35,6 +36,11 @@ function ItemName({
   dashboardId: DashboardId;
   type?: "collections" | "snippet-collections";
 }) {
+  const { data: collection } = useGetCollectionQuery(
```
This won't fire on root collection, but maybe that's alright, since it will still get the standard `collection` icon

npfitz commented on frontend/src/metabase/common/components/ConfirmModal/ConfirmModal.tsx:
```diff
@@ -51,6 +53,7 @@ export const ConfirmModal = ({
       <Flex direction="column" gap="lg" mt="lg">
         {content ? <Text>{content}</Text> : null}
         <Text>{message}</Text>
+        {children}
```
Do we need `content`, `message`, *and* `children` props here? Can we at least reduce it to 2 props?

npfitz commented on frontend/src/metabase/lib/urls/modelToUrl.ts:
```diff
@@ -1,9 +1,12 @@
+import type { CollectionId } from "metabase-types/api";
```
unused

edpaget commented on enterprise/backend/src/metabase_enterprise/remote_sync/api.clj:
```diff
@@ -0,0 +1,279 @@
+            (remote-sync.task/create-sync-task! ttype api/*current-user-id*)))]
+    (api/check-400 (not existing?) "Remote sync in progress")
+    (u.jvm/in-virtual-thread*
+     (dh/with-timeout {:interrupt? true
```
Maybe it would be better to make the timeout here a multiple of the remote-sync-task-time-limit-ms instead? I'd like this setting to default to a lower value now that progress is regularly reported so this is a safety-valve and doesn't need to following this setting exactly.

npfitz commented on frontend/src/metabase-types/api/search.ts:
```diff
@@ -88,6 +88,7 @@ export interface SearchResult<
   pk_ref: FieldReference | null;
   table_schema: string | null;
   collection_authority_level: "official" | null;
+  type?: "instance-analytics" | "trash" | "remote-synced" | null;
```
should this pull in `CollectionType`? Also, is this only for collections? if so, maybe the prop should be named to reflect that

npfitz reviewed (commented):
First pass on FE code. Purely just looking at the code itself, not the UX behind it

edpaget commented on enterprise/backend/src/metabase_enterprise/remote_sync/api.clj:
```diff
@@ -0,0 +1,279 @@
+(defn- async-export!
+  [branch message]
+  (let [source (source/source-from-settings branch)
+        last-task-version (remote-sync.task/last-version)
```
Should this be a concern of the impl instead of happening here? I would like to expose fewer implementation details of the `source` module here. That means the API just concered with the plumping of running the task and the impl handles all the lower level details.

npfitz commented on enterprise/frontend/src/metabase-enterprise/remote_sync/components/CollectionSyncStatusBadge.tsx:
```diff
@@ -0,0 +1,29 @@
+
+  return (
+    <Tooltip label={t`Unsynced changes`}>
+      <Box bdrs="50%" bg="warning" h={12} w={12} mr="xs" />
```
Oh geez... looks like my `inline code` didn't come through VS code. 

```
Should this be an `Indicator` or an `Icon`
```

edpaget commented on enterprise/backend/src/metabase_enterprise/remote_sync/api.clj:
```diff
@@ -0,0 +1,279 @@
+  1. Fetch the latest changes from the configured git repository
+  2. Load the updated content using the serialization/deserialization system
+
+  If force=false (default) and there are unsaved changes in the Remote Sync collection, the import return a 400 response.
```
```suggestion
  If force=false (default) and there are unsaved changes in the Remote Sync collection, the import action will return a 400 response.
```

edpaget commented on enterprise/backend/src/metabase_enterprise/remote_sync/api.clj:
```diff
@@ -0,0 +1,279 @@
+  {:is_dirty (remote-sync.object/dirty-global?)})
+
+(api.macros/defendpoint :get "/dirty"
+  "Return dirty models from a any remote-sync collection"
```
```suggestion
  "Return dirty models from any remote-synced collection"
```

edpaget commented on enterprise/backend/src/metabase_enterprise/remote_sync/api.clj:
```diff
@@ -0,0 +1,279 @@
+     :message (when-not task-id "No changes since last import")}))
+
+(api.macros/defendpoint :get "/is-dirty"
+  "Check if any collection has remote sync changes that are not saved."
```
```suggestion
  "Check if any remote-synced collection has remote sync changes that are not saved."
```

edpaget commented on enterprise/backend/src/metabase_enterprise/remote_sync/api.clj:
```diff
@@ -0,0 +1,279 @@
+   {:keys [new_branch message]} :- [:map
+                                    [:new_branch ms/NonBlankString]
+                                    [:message ms/NonBlankString]]]
+  (api/check-superuser)
```
Is this still the right permission?

edpaget commented on enterprise/backend/src/metabase_enterprise/remote_sync/core.clj:
```diff
@@ -0,0 +1,24 @@
+
+(defenterprise editable?
+  "Should the remote-synced collection be editable. Always true on OSS"
+  :feature :none
```
```suggestion
  :feature :remote-sync
```

edpaget commented on enterprise/backend/src/metabase_enterprise/remote_sync/impl.clj:
```diff
@@ -0,0 +1,174 @@
+(defn- affected-collections
+  "Get all collections that are descendants of those in the sync and the ones synced"
+  []
+  ;; Otherwise get all remote-synced collections
```
```suggestion
```

edpaget commented on enterprise/backend/src/metabase_enterprise/remote_sync/impl.clj:
```diff
@@ -0,0 +1,174 @@
+    (do
+      (log/errorf e "Failed to reload from git repository: %s" (ex-message e))
+      (analytics/inc! :metabase-remote-sync/imports-failed)
+      (let [error-msg (cond
```
I think this is the same error handling that we have in the `/branches` API let's pull this into a separate function for reuse.

edpaget commented on enterprise/backend/src/metabase_enterprise/remote_sync/settings.clj:
```diff
@@ -0,0 +1,105 @@
+  :export? false)
+
+(defsetting remote-sync-token
+  (deferred-tru "A GH token")
```
```suggestion
  (deferred-tru "An Authorization Bearer token allowing access to the git repo over HTTP")
```

edpaget commented on enterprise/backend/src/metabase_enterprise/remote_sync/settings.clj:
```diff
@@ -0,0 +1,105 @@
+  :encryption :no
+  :default (* 1000 60 60))
+
+(defn check-git-settings
```
Does this need to be public?

nvoxland replied:
Currently not, since it's just used in `check-and-update-remote-settings!`. If we ever want to check it without setting it we could make it public then?

edpaget commented on enterprise/backend/src/metabase_enterprise/remote_sync/source/git.clj:
```diff
@@ -0,0 +1,289 @@
+                        analytics-labels e))))))
+
+(defn- call-remote-command [^TransportCommand command {:keys [^String token]}]
+  (let [analytics-labels {:operation (-> command .getClass .getName) :remote true}
```
Maybe a `class-name` helper fn

edpaget commented on enterprise/backend/src/metabase_enterprise/remote_sync/source/git.clj:
```diff
@@ -0,0 +1,289 @@
+  "Clone git repository to a temporary directory using jgit. If the temp directory already exists, use it rather than re-clone.
+  Returns the path to the cloned `jgit.Repository` bare-repo object."
+  [{:keys [^String url ^String token] :as args}]
+  (let [dir (io/file (System/getProperty "java.io.tmpdir") "metabase-git" (-> (str/join ":" [url token]) buddy-hash/sha1 codecs/bytes->hex))
```
Since the token is sensitive should we use a stronger hash here than sha1? Obviously it should be difficult to reverse it since a token is already random, but want to make sure we don't leak it. 

Slash it would maybe be better to not use the token for this?

edpaget commented on enterprise/backend/src/metabase_enterprise/remote_sync/source/git.clj:
```diff
@@ -0,0 +1,289 @@
+                                          (.setBare true)) {:token token})
+          (log/info "Successfully cloned repository" {:dir dir}))
+        (catch Exception e
+          (throw (ex-info (format "Failed to clone git repository: %s" (.getMessage e))
```
```suggestion
          (throw (ex-info (format "Failed to clone git repository: %s" (ex-message e))
```

edpaget commented on enterprise/backend/src/metabase_enterprise/remote_sync/source/git.clj:
```diff
@@ -0,0 +1,289 @@
+      (let [log-result (call-command (-> (.log git)
+                                         (.add branch-id)))]
+        (map (fn [^RevCommit commit] {:message      (.getFullMessage commit)
+                                      :author-name  (.getName (.getAuthorIdent commit))
```
small nit, i prefer the `..` interop macro for chains of calls like this so
```suggestion
                                      :author-name  (.. commit getAuthorIdent getName)
```
just a bit easier to read. imo, some other examples in this file.

edpaget commented on enterprise/backend/src/metabase_enterprise/remote_sync/source/ingestable.clj:
```diff
@@ -0,0 +1,122 @@
+(defn ingestable-version
+  "Returns the version of the given ingestable source."
+  [ingestable]
+  (let [ingestable (or (:ingestable ingestable) ingestable)]
```
This function feels weird to me, maybe we should have a `source` object available whereever we need a version instead of trying to unwrap it this way?

johnswanson replied:
+1 on that

edpaget commented on enterprise/backend/src/metabase_enterprise/remote_sync/task/table_cleanup.clj:
```diff
@@ -0,0 +1,57 @@
+                 (triggers/with-identity cleanup-trigger-key)
+                 (triggers/start-now)
+                 (triggers/with-schedule
+                   ;; Run daily at 2:19 AM
```
```suggestion
                   ;; Run daily at 2:29 AM
```

andsalves committed:
Add options modal when trying to push while remote is ahead

edpaget commented on enterprise/backend/test/metabase_enterprise/remote_sync/models/remote_sync_task_test.clj:
```diff
@@ -0,0 +1,402 @@
+  (f)
+  (t2/delete! :model/RemoteSyncTask))
+
+(use-fixtures :each cleanup-sync-tasks)
```
Replace with the cleaning fn in test-helper

edpaget commented on enterprise/backend/test/metabase_enterprise/remote_sync/permissions_test.clj:
```diff
@@ -0,0 +1,270 @@
+(deftest remote-synced-permissions-with-remote-sync-type-import-test
+  (testing "can_write should be false for remote-synced collection items when remote-sync-type is production"
+    (mt/with-current-user (mt/user->id :rasta)
+      (mt/with-temporary-setting-values [settings/remote-sync-type :production]
```
with-temporary-setting-values prevents us from using `^:parallel`

edpaget committed:
Support force option on export

edpaget committed:
Update branch setting and complete task in txn

edpaget commented on enterprise/backend/test/metabase_enterprise/remote_sync/source/ingestable_test.clj:
```diff
@@ -0,0 +1,146 @@
+              (let [final-task (t2/select-one :model/RemoteSyncTask :id task-id)]
+                (is (< (abs (- (:progress final-task) (double normalize))) 0.01)
+                    "Progress should equal normalize value when all items ingested")))))
+        (finally
```
use mt/with-model-cleanup instead.

edpaget commented on enterprise/backend/test/metabase_enterprise/remote_sync/source_test.clj:
```diff
@@ -0,0 +1,152 @@
+        (testing "files list is empty"
+          (is (empty? (:files @written-files))))))))
+
+(deftest store!-large-stream-test
```
this is a pointless test.

edpaget commented on src/metabase/collections/api.clj:
```diff
@@ -1381,12 +1399,24 @@
     (move-or-archive-collection-if-needed! collection-before-update collection-updates)
-    (events/publish-event! :event/collection-touch {:collection-id id :user-id api/*current-user-id*}))
+    (let [updated-collection (t2/select-one :model/Collection :id id)]
+      (when config/ee-available?
```
is there a better way to do this? we have to because if a public-event! method isn't defined this will error.

edpaget committed:
Merge branch 'master' into source-of-truth

edpaget committed:
Let export! set branch in stash and fix tests

edpaget committed:
Fix PR feedback nits

edpaget committed:
Update enterprise/backend/src/metabase_enterprise/remote_sync/api.clj

edpaget committed:
Consistent documentation style in remote-sync

andsalves committed:
Add option to pull from remote / Adjust push/pull icons / Adjust out-of-sync options modal

andsalves committed:
modals UI polish

andsalves committed:
constants clean up

edpaget committed:
Add api schemas and make error messages consistent

johnswanson commented on enterprise/backend/src/metabase_enterprise/documents/models/document.clj:
```diff
@@ -43,11 +42,10 @@
   [old-collection-id new-collection-id]
   (when old-collection-id
-    (collection/check-write-perms-for-collection old-collection-id))
+    (api/check-403 (mi/can-write? (t2/select-one :model/Collection :id old-collection-id))))
```
I think you can just do `(api/read-check :model/Collection old-collection-id)`

johnswanson commented on enterprise/backend/src/metabase_enterprise/remote_sync/task/table_cleanup.clj:
```diff
@@ -0,0 +1,57 @@
+    (log/infof "Deleted %d remote_sync_task records. Cleanup successful." deleted-count)
+    deleted-count))
+
+(defn- trim-tables!
```
why have 2 functions for this when they do the same thing?

edpaget replied:
Need to refactor there used to be two tables that needed to be truncated.

johnswanson commented on enterprise/backend/src/metabase_enterprise/remote_sync/source/git.clj:
```diff
@@ -0,0 +1,289 @@
+        (let [tree-id (.writeTree index inserter)
+              commit-builder (doto (CommitBuilder.)
+                               (.setTreeId tree-id)
+                               (.setAuthor (PersonIdent. "Metabase Library" "library@metabase.com"))
```
if we do this we should inform the user that it'll be attributed to them, or give them the option - someone might not want the email they use with Metabase to be on GH.

johnswanson reviewed (commented):
I can't approve because I opened this PR originally 😂 but this LGTM, some small suggestions

edpaget commented on enterprise/backend/src/metabase_enterprise/documents/api/document.clj:
```diff
@@ -142,43 +143,44 @@
+                                    :content_type prose-mirror/prose-mirror-content-type}
+                                   cards-to-update-in-ast)))
+                    (u/prog1 (get-document document-id)
+                      (when (collections/remote-synced-collection? (:collection_id <>))
```
I think the answer is no from talking with @nvoxland

edpaget committed:
fix check-403 to write-check

edpaget committed:
Merge branch 'master' into source-of-truth

andsalves committed:
Clean up a couple components / Move things around

edpaget committed:
Apply suggestions from code review

Co-authored-by: metabase-bot[bot] <109303359+metabase-bot[bot]@users.noreply.github.com>

npfitz committed:
temp

edpaget committed:
Address renaming suggestions

edpaget committed:
Fix var names and bad merge

edpaget committed:
Merge branch 'master' into source-of-truth

edpaget committed:
double->float in schema

edpaget committed:
Share error handling

edpaget committed:
Update enterprise/backend/src/metabase_enterprise/remote_sync/settings.clj

edpaget committed:
Update enterprise/backend/src/metabase_enterprise/remote_sync/settings.clj

edpaget committed:
Update enterprise/backend/src/metabase_enterprise/remote_sync/settings.clj

edpaget committed:
Addressing small PR comments

andsalves committed:
Refetch properties when branch is created (current branch changed)

andsalves committed:
Make sure /current-task is only called when import and export are not in progress

andsalves committed:
Replace unset color variables

andsalves committed:
Improve element padding on side panel

andsalves committed:
Remove a couple trailing spaces reported by clj whitespace linter

andsalves committed:
Show error when settings request fails / Make sure current-task also waits for the PUT settings mutation

edpaget committed:
Fix document perm tests

nvoxland committed:
Allow empty remote-url, moved import and export calls to impl

nvoxland committed:
Merge remote-tracking branch 'origin/source-of-truth' into source-of-truth

andsalves committed:
Fix minor prettier issue

andsalves committed:
Merge branch 'master' into source-of-truth

andsalves committed:
Fix another lil lint issue

npfitz committed:
e2e push / pull

edpaget committed:
Unify more of the import implementation

edpaget committed:
fix CI failures

edpaget committed:
fix mysql ci failures

edpaget committed:
Address metabot critiques

edpaget committed:
Merge branch 'master' into source-of-truth

edpaget committed:
Address additional metabot feedback

edpaget committed:
last round of docstring feedback

npfitz committed:
library admin settings e2e

edpaget committed:
Add attribution to TODO

edpaget committed:
fix ws

edpaget committed:
fix missed commiting file

edpaget committed:
fix another missing file

npfitz committed:
production switch branches

edpaget committed:
More clojure like docstrings

edpaget committed:
More docstring fixes

npfitz committed:
adding branch switching test

andsalves committed:
Address Nick's PR review comments (most of them)

edpaget committed:
Merge branch 'master' into source-of-truth

edpaget committed:
Fix more nits

andsalves committed:
Resolving Nick's comments part 2

nvoxland committed:
Improving remote-sync settings handling

nvoxland committed:
Merge remote-tracking branch 'origin/source-of-truth' into source-of-truth

mazameli committed:
attempt to repair synced_folder icon

mazameli committed:
Merge branch 'source-of-truth' of github.com:metabase/metabase into source-of-truth

nvoxland committed:
Fail if attempting to connect to an uninitialized repo

nvoxland committed:
Merge remote-tracking branch 'origin/source-of-truth' into source-of-truth

mazameli committed:
correct the name of synced_collection.svg

mazameli committed:
Merge branch 'source-of-truth' of github.com:metabase/metabase into source-of-truth

mazameli committed:
add new repository icon

edpaget committed:
Add default-branch to source protocol

eliotmetabase committed:
Update Git Sync to Remote Sync in Admin settings (#64993)

edpaget committed:
Make api docstrings and errors more generic

edpaget committed:
Use check-400 to simplify api

edpaget committed:
fix ws

andsalves committed:
Rollback BaseSMTPConnectionForm.tsx (fixes a e2e test)

npfitz committed:
fix e2e tests?

edpaget committed:
Merge branch 'master' into source-of-truth

andsalves committed:
Fix small merge issue

nvoxland committed:
Always use `last-version` instead of caring about import vs. export version

nvoxland committed:
Merge remote-tracking branch 'origin/source-of-truth' into source-of-truth

npfitz committed:
starting warning modal tests

nvoxland commented on enterprise/backend/src/metabase_enterprise/remote_sync/source/git.clj:
```diff
@@ -0,0 +1,289 @@
+  [{:keys [^Git git] :as git-source}]
+  (log/info "Fetching repository" {:repo (str git)})
+  (u/prog1 (call-remote-command (.fetch git) git-source))
+  (log/info "Successfully fetched repository" {:repo (str git)}))
```
Took it out of the message, just in case

edpaget committed:
Fix broken test

nvoxland committed:
Fetch source at the beginning of import/export operations

nvoxland committed:
Delete branches that were deleted on remote

nvoxland committed:
Merge remote-tracking branch 'origin/source-of-truth' into source-of-truth

edpaget committed:
Fix missing test suffix

edpaget committed:
Fix eastwood

edpaget committed:
Merge branch 'master' into source-of-truth

andsalves committed:
Adjust branching e2e test, skip failing tests
